### PR TITLE
python: release GIL around hy_goal_run* calls

### PR DESCRIPTION
**Pull request by mizdebsk**
Related to commit 745dfd4.  Depsolving can be a time-consuming
operation and it should not prevent other Python threads from running.

### CODE REVIEW

**Comment by jsilhan** [comment]
Depsolving takes ~ tenth of second. Do you experience huge speed-up? Can you share the benchmarks, please?

**Comment by mizdebsk** [comment]
Yes, speed-up is really big, at least for my use case (running hundreds of goals in different threads).

In my test setup I have 4 threads running 10,000 goals each, in pararell. Without the patch, average time to run a single goal is about **199 ms**, while with the patch applied it's **72 ms**. In this case the patch improved performance 2.76 times. With more threads you can get even more speedup.

Benchmark code used:
```
#!/usr/bin/python

import hawkey
import time
from concurrent.futures import ThreadPoolExecutor

n_sacks = 4
n_goals = 10000

sack_time = 0
goal_time = 0

def benchmark():
    global sack_time
    global goal_time
    sack_time = sack_time - time.time()
    sack = hawkey.Sack(arch='x86_64')
    for arch in ['i386', 'x86_64']:
        repo = hawkey.Repo(arch)
        repo.repomd_fn = arch + '-repomd.xml'
        repo.primary_fn = arch + '-primary.xml.gz'
        repo.filelists_fn = arch + '-filelists.xml.gz'
        sack.load_yum_repo(repo, load_filelists=True)
    sack_time = sack_time + time.time()

    goal_time = goal_time - time.time()
    for i in xrange(0, n_goals):
        goal = hawkey.Goal(sack)
        problems = []
        for name in ['eclipse-pde', 'gnome-shell', 'texlive']:
            goal.install(select=hawkey.Selector(sack).set(name=name))
        goal.run()
    goal_time = goal_time + time.time()

with ThreadPoolExecutor(max_workers=n_sacks) as executor:
    [f.result() for f in [executor.submit(benchmark) for i in xrange(0,n_sacks)]]
        

print "Average time to load sack: " + str(sack_time / n_sacks)
print "Average time to run goal : " + str(goal_time / n_sacks / n_goals)
```
Results without this patch:
```
Average time to load sack: 30.3952310681
Average time to run goal : 0.0198605083048
```
Results with this patch:
```
Average time to load sack: 30.8735132813
Average time to run goal : 0.00721658264399
```

File: src/python/goal-py.c [hunk 1]
```diff
@@ -411,8 +414,10 @@ run_all(_GoalObject *self, PyObject *args, PyObject *kwds)
 	return NULL;
 
     struct _PySolutionCallback cb_s = {callback_tuple, callback, 0};
-    int ret = hy_goal_run_all_flags(self->goal, py_solver_callback, &cb_s,
-				    flags);
+    int ret;
+    Py_BEGIN_ALLOW_THREADS;
```
Comment (jsilhan) [inline src/python/goal-py.c]: this causes tests to fail. It takes python function callback and segfaults in hawkey python tests.

Comment (mizdebsk) [inline src/python/goal-py.c]: > this causes tests to fail. It takes python function callback and segfaults in hawkey python tests.

Fixed in bbc18e1

**Comment by jsilhan** [comment]
Jenkins, test this please

**Comment by j-mracek** [comment]
Jenkins, test this please

**Comment by Conan-Kudo** [comment]
@mizdebsk You may want to adapt this for [libhif](https://github.com/rpm-software-management/libhif) and submit a PR there, as the hawkey code has been merged into libhif.

**Comment by jsilhan** [comment]
Merged, there. Thanks.

**Pull request closed without merging**